### PR TITLE
Flow metrics and events

### PR DESCRIPTION
**Pull request by alakendu**
This enhancement will address https://github.com/open-traffic-generator/models/pull/108

### CODE REVIEW

File: snappi_ixnetwork/trafficitem.py [hunk 1]
```diff
@@ -422,26 +477,25 @@ def _configure_pattern(self,
...
-            ixn_field.update(Auto=False,
-                            ValueType='decrement',
-                            ActiveFieldChoice=field_choice,
-                            StartValue=pattern.decrement.start,
-                            StepValue=pattern.decrement.step,
-                            CountValue=pattern.decrement.count)
-            # except Exception as e:
-            #     print(e)
+            try:
+                ixn_field.update(Auto=False,
```
Comment (ANISH-GOTTAPU) [inline snappi_ixnetwork/trafficitem.py:491]: `test_errors.py` test is failing in CI/CD as try, except block is bypassing the server side error.

Comment (alakendu) [reply snappi_ixnetwork/trafficitem.py]: fix in https://github.com/open-traffic-generator/ixnetwork/commit/04c5079dbc50aa2d3eecfc840ab95371e0298f65

File: snappi_ixnetwork/trafficitem.py [hunk 2]
```diff
@@ -690,10 +738,26 @@ def results(self, request):
             msg = "Invalid format of flow_names passed {},\
                     expected list".format(flow_names)
             raise Exception(msg)
-
-        filter = {'property': 'name', 'regex': '.*'}
-        filter['regex'] = '^(%s)$' % '|'.join(
-                        self._api.special_char(flow_names))
+        final_flow_names = [
+            f.name for f in self._api._config.flows
+            if f.metrics.enable is True and f.name in flow_names
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:743]: shouldn't we handle the case where `f.metrics` could be None ?

File: snappi_ixnetwork/trafficitem.py [hunk 3]
```diff
@@ -690,10 +738,26 @@ def results(self, request):
...
+        ]
+        if final_flow_names == []:
+            msg = """
+            To fetch flow metrics at least one flow shall have metric enabled
+            """
+            raise Exception(msg.strip())
+        diff = set(flow_names).difference(final_flow_names)
+        if len(diff) > 0:
+            self._api.warning(
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:752]: We shouldn't have to throw warnings here.
- If flow_names=[], we should only return flow metrics for flows where metrics is enabled
- If flow_names contains name of at least one flow where metrics is not enabled, we should raise error

Error message should say:
`Flow metrics is not enabled on flow %s`

File: snappi_ixnetwork/trafficitem.py [hunk 4]
```diff
@@ -720,14 +784,34 @@ def results(self, request):
                         )
                     flow_rows[traffic_item['name'] + stream['txPortName'] + rx_port_name] = flow_row
 
-        # resolve result values
-        table = self._api.assistant.StatViewAssistant(
+        flows_has_latency = []
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:787]: this whole block should be done as part of set_config

File: snappi_ixnetwork/trafficitem.py [hunk 5]
```diff
@@ -743,5 +827,48 @@ def results(self, request):
...
+            latency_map = TrafficItem._RESULT_LATENCY_CUT_THROUGH
+        latency_result = {}
+        for external_name, internal_name, external_type in latency_map:
+            if internal_name not in row.Columns:
+                continue
+            try:
+                self._set_result_value(latency_result, external_name,
+                                       row[internal_name], external_type)
+            except Exception:
+                self._api.warning("Issue at the time processing latency result")
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:851]: raise an error: `Could not fetch latency metrics: %s` % exception_err

File: snappi_ixnetwork/trafficitem.py [hunk 6]
```diff
@@ -743,5 +827,48 @@ def results(self, request):
...
+                val = row[internal_name].split(':')
+                mul = [3600, 60, 1]
+                sv = sum([
+                    int(float(v) * 10 ** 9 + 0.1) * mul[i]
+                    for i, v in enumerate(val)
+                ])
+                self._set_result_value(timestamp_result, external_name,
+                                       sv, external_type)
+            except Exception:
+                self._api.warning("Issue at the time processing timestamp result")
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:871]: raise an error: `Could not fetch timestamp metrics: %s` % exception_err

File: snappi_ixnetwork/trafficitem.py [hunk 7]
```diff
@@ -743,5 +827,48 @@ def results(self, request):
...
+                self._api.warning("Issue at the time processing latency result")
+                pass
+        if len(latency_result) > 0:
+            flow_row['latency'] = latency_result
+    
+    def _construct_timestamp(self, flow_row, row):
+        timestamp_result = {}
+        for external_name, internal_name, external_type in TrafficItem._RESULT_TIMESTAMP:
+            if internal_name not in row.Columns:
+                continue
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:860]: should raise an error here as well

File: snappi_ixnetwork/trafficitem.py [hunk 8]
```diff
@@ -743,5 +827,48 @@ def results(self, request):
...
+
+    def _construct_latency(self, flow_row, latency_mode, row):
+        if latency_mode == 'store_forward':
+            latency_map = TrafficItem._RESULT_LATENCY_STORE_FORWARD
+        else:
+            latency_map = TrafficItem._RESULT_LATENCY_CUT_THROUGH
+        latency_result = {}
+        for external_name, internal_name, external_type in latency_map:
+            if internal_name not in row.Columns:
+                continue
```
Comment (ashutshkumr) [inline snappi_ixnetwork/trafficitem.py:846]: should raise an error here as well

**Review by ashutshkumr — CHANGES_REQUESTED** [review]
@ANISH-GOTTAPU please add new unit TC (one should do) at least for following scenarios:
- enable metrics on some flows and disable on some; fetching with flow_names=[] should return only valid flow metrics; fetching flow_names containing mix of flows should raise error.
- try fetching timestamps / latency if metrics is enabled but timestamp/latency is not.
- using different latency modes should raise warning

File: snappi_ixnetwork/trafficitem.py [hunk 9]
```diff
@@ -200,15 +222,48 @@ def config(self):
...
+    def _process_latency(self, latency, config_latency):
+        if config_latency is None:
+            if latency.mode is not None:
+                config_latency = latency.mode
+            else:
+                config_latency = 'store_forward'
+        else:
+            if latency.mode is not None and \
+                    config_latency != latency.mode:
+                raise Exception("Solution only accept same type of latency mode")
```
[the remark this answers is not in the document]
Comment (alakendu) [reply snappi_ixnetwork/trafficitem.py:251]: Thanks!

File: snappi_ixnetwork/trafficitem.py (hunk 2 above)
Comment (alakendu) [reply snappi_ixnetwork/trafficitem.py:743]: done

**Pull request merged**